File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        ret = []
        if not root:
            return ret
        q = [(root, 0)]
        prevDepth = 0
        prevElem = root
        while q:
            curr, depth = q.pop(0)
            if depth != prevDepth:
                prevDepth = depth
                ret.append(prevElem.val)
            prevElem = curr 
            if curr.left:
                q.append((curr.left, depth + 1))
            if curr.right:
                q.append((curr.right, depth + 1))
        ret.append(prevElem.val)
        return ret

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root)
            return {};
        auto q = queue<TreeNode*>({root});
        vector<int> res2;
        while (q.size()) {
            int levelSize = q.size();
            int currVal = -1;
            for (int i = 0; i < levelSize; i++) {
                auto curr = q.front();
                q.pop();
                currVal = curr->val;
                if (curr->left)
                    q.push(curr->left);
                if (curr->right)
                    q.push(curr->right);
            }
            res2.push_back(currVal);
        }
        return res2;
    }
};
